custom_components/google_home/api.py

```diff
@@ -313,7 +314,66 @@ async def delete_alarm_or_timer(
                 device.name,
             )
         except ClientError as ex:
-            # Make sure that we log the exception if one occurred.
+            # Make sure that we log the exception from the client if one occurred.
+            _LOGGER.error(
+                "Request error: %s",
+                ex,
+            )
+
+    async def reboot_google_device(self, device: GoogleHomeDevice) -> None:
+        """Reboots a Google Home device if supported."""
+
+        if device.ip_address is None:
+            _LOGGER.warning("Device %s doesn't have IP address!", device.name)
+            return
+
+        if device.auth_token is None:
+            _LOGGER.warning("Device %s doesn't have a auth token!", device.name)
+            return
+
+        url = self.create_url(device.ip_address, PORT, API_ENDPOINT_REBOOT)
+
+        # We need to remove charset=UTF-8 or else it will return a 400 Bad Request.
+        HEADERS.update(
+            {
+                HEADER_CAST_LOCAL_AUTH: device.auth_token,
+                HEADER_CONTENT_TYPE: "application/json",
+            }
+        )
+
+        # "now" means reboot and "fdr" means factory reset (Not implemented).
+        data = {"params": "now"}
+
+        _LOGGER.debug(
+            "Trying to reboot Google Home device %s - %s",
+            device.name,
+            url,
+        )
+
+        try:
+            async with self._session.post(
+                url, json=data, headers=HEADERS, timeout=TIMEOUT
+            ) as response:
+                if response.status == HTTP_OK:
+                    # It will return 200 even if the device does not support rebooting.
+                    _LOGGER.info(
+                        "Successfully asked %s to reboot.",
+                        device.name,
+                    )
+                else:
+                    _LOGGER.error(
+                        "Failed to reboot %s, API returned" " %d: %s",
+                        device.name,
+                        response.status,
+                        response,
+                    )
+        except ClientConnectorError:
+            _LOGGER.warning(
+                "Failed to connect to %s device. The device is probably offline.",
+                device.name,
+            )
+        except ClientError as ex:
+            # Make sure that we log the exception from the client if one occurred.
```

Comment:
These services have a lot of duplicated code. Can it be abstracted with a common function?

Reply:
Yes, i thought that too, i even have another thing planned. Should i refactor it in this or should we move it to another PR?

Reply:
I'd rather have it here as this adds a new copy.

Reply:
Not sure, if it is really worth it. I want to start replacing api.py with googledevices, once I have finished up another project. While I agree, that some amount de-duplication would likely be beneficial, we should not invest too much here.

Reply:
Well, I think in any case we should try to keep the code here good. I don't think this deduplication should be very hard.

Reply:
It is fine, i am nearly done with it. We/you can also use this when making `googledevices`. So I don't see it as a total waste of time. 😄 I am also gonna add a "Do not disturb switch" soon.